algorithms/kernel/dtrees/forest/oneapi/df_feature_type_helper_oneapi.i

```diff
@@ -0,0 +1,567 @@
+
+    {
+        KernelArguments args(3);
+        args.set(0, values, AccessModeIds::read);
+        args.set(1, binOffsets, AccessModeIds::read);
+        args.set(2, binBorders, AccessModeIds::write);
+
+        KernelRange global_range(maxBins);
+
+        context.run(global_range, kernel, args, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+    }
+
+    return status;
+}
+
+template <typename algorithmFPType>
+services::Status IndexedFeaturesOneAPI<algorithmFPType>::computeBins(UniversalBuffer & values, UniversalBuffer & indices,
+                                                                     UniversalBuffer & binBorders, UniversalBuffer & bins, int nRows, int nBins,
+                                                                     int localSize, int nLocalBlocks)
+{
+    DAAL_ITTNOTIFY_SCOPED_TASK(indexedFeatures.computeBins);
+
+    services::Status status;
+
+    auto & context = services::Environment::getInstance()->getDefaultExecutionContext();
+    auto & factory = context.getClKernelFactory();
+    __buildProgram<algorithmFPType>(factory);
+
+    auto kernel = factory.getKernel("computeBins");
+
+    {
+        KernelArguments args(6);
+        args.set(0, values, AccessModeIds::read);
+        args.set(1, indices, AccessModeIds::read);
+        args.set(2, binBorders, AccessModeIds::read);
+        args.set(3, bins, AccessModeIds::write);
+        args.set(4, nRows);
+        args.set(5, nBins);
+
+        KernelRange local_range(localSize);
+        KernelRange global_range(localSize * nLocalBlocks);
+
+        KernelNDRange range(1);
+        range.global(global_range, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+        range.local(local_range, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+
+        context.run(range, kernel, args, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+    }
+
+    return status;
+}
+
+template <typename algorithmFPType>
+services::Status IndexedFeaturesOneAPI<algorithmFPType>::computeBins(UniversalBuffer & values, UniversalBuffer & indices, UniversalBuffer & bins,
+                                                                     FeatureEntry & entry, int nRows, const dtrees::internal::BinParams * pBinPrm)
+{
+    services::Status status;
+
+    auto & context = services::Environment::getInstance()->getDefaultExecutionContext();
+
+    const int maxBins      = pBinPrm->maxBins < nRows ? pBinPrm->maxBins : nRows;
+    const int localSize    = _preferableSubGroup;
+    const int nLocalBlocks = 1024 * localSize < nRows ? 1024 : (nRows / localSize) + !!(nRows % localSize);
+
+    auto binOffsets = context.allocate(TypeIds::id<int>(), maxBins, &status);
+    auto binBorders = context.allocate(TypeIds::id<algorithmFPType>(), maxBins, &status);
+
+    DAAL_CHECK_STATUS_VAR(status);
+
+    {
+        auto binOffsetsHost = binOffsets.template get<int>().toHost(ReadWriteMode::writeOnly);
+        int offset          = 0;
+        for (int i = 0; i < maxBins; i++)
```

Comment:
It looks like it's possible to implement the cycle on GPU using inclusive scan.

Reply:
thanks, i will try it on next optimization iteration (not for this check in)